### DISCUSSION TITLE
How to detect unmerged document changes.

### DISCUSSION THREAD

**Discussion by kfertitta** [dotnet/roslyn#76236] [Q&A] [answered]
We have refactorings that operate on multi-targeted projects. We'd like to detect if a series of syntax operations conducted on more than one TFM has produced unmerged document changes before calling `TryApplyChanges`. This seems to almost always represent a bug in our own logic whereby we're editing one TFM in a place that overlaps with an edit in another TFM (and a quick look at the Roslyn source seems to indicate these conflicts are, in fact, produced when TFM edits overlap in spans). I'd like to have `TryApplyChanges` fail or have a way to detect the presence of any of these unmerged document changes so that I can rollback the entire series of operations rather than have Roslyn inject the unmerged change comments.

Thanks in advance.

**Comment by CyrusNajmabadi** [comment]
Refactorings are given the entire solution snapshot to start with. And they produce the entire solution snapshot they end with. 

So you can just compare the start and end snapshot to get the changes. Then for any limited files see if the changes are not compatible and revert those from those documents.

**Comment by kfertitta** [comment]
Thanks @CyrusNajmabadi .  I think the bit I'm not seeing is the last step whereby I check for compatibility of document changes. The Roslyn source I saw seemed to only manifest these overlapping edits by injecting comments, though I obviously could be overlooking something. Was hoping for some higher level API that lets me check for the compatibility of specific docs, or do I have to replicate the span-overlap-searching code in Roslyn?  I've looping thru the diff's like so:

```
            var changes = workspace.CurrentSolution.GetChanges(newSolution);

            var projectChanges = changes.GetProjectChanges().ToList();

            foreach (var projectChange in projectChanges)
            {
                var docIds = projectChange.GetChangedDocuments();
            }
```

**Reply by CyrusNajmabadi** [reply, marked as answer]
> or do I have to replicate the span-overlap-searching code in Roslyn

I think you'll need to do that. When you have multiple related document changes, just get all the changes for each document in the related DOC group. Then see if any edit spans overlap but are not identical to what the other documents have.

**Reply by CyrusNajmabadi** [reply]
> I've looping thru the diff's like so:

No need for ToList. 

Also, use GetRelatedXxx to find the corresponding linked document IDs in other projects.

**Reply by kfertitta** [reply]
That all sounds reasonable and doable.  Thanks much for the quick guidance.
